restorer: allow writing target file blobs out of order

ifedorenko opened the pull request:
What is the purpose of this change? What does it change?
--------------------------------------------------------

Much simpler implementation that guarantees each required pack
is downloaded only once (and hence does not need to manage
pack cache). Also improves large file restore performance.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

See #2074
https://forum.restic.net/t/restore-using-rclone-gdrive-backend-is-slow/1112/8
https://forum.restic.net/t/degraded-restore-performance-s3-backend/1400

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have added tests for all changes in this PR
- [ ] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [ ] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [x] I'm done, this Pull Request is ready for review

ifedorenko commented:
@fd0 @mholt can you review this PR when you get a chance? It is the code discussed in #2074, and should be ready or almost ready to be merged to master.

Compared to master, this PR does not guarantee target files are written sequentially, start to finish, which allowed much simpler implementation _and_ faster multithreaded restore of very large files.

mholt commented:
Sure -- I just barely finished my masters thesis so I'm in major catch-up mode, but at a glance I _really_ like the ~1K LoC reduction. Don't wait up on me, but I will try to give it a high-level once over soon.

DurvalMenezes commented:
Hello everyone, and specially @ifedorenko.

Just to make sure: this PR objective is *speed* optimization, **not** memory usage, right? 

I ask because [it's been implied in the forum](https://forum.restic.net/t/restic-memory-use-during-restore-v0-9-1-vs-0-9-5/1742/2) that it could help me save some memory during the restore, but I'm seeing the exact opposite (about *double* memory usage as compared to "standard" v0.9.5 restic... 

PS: if you need anything, I compiled it with `--tags profile`, so I can provide anything the pprof runtinme webserver makes available. 

Cheers, 
-- Durval.

ifedorenko commented:
This PR is supposed to use little less memory compared to 0.9.5, can't think of why it takes 2x memory instead. I'd need to know what data structures use the memory in your tests to suggest anything specific.

DurvalMenezes commented:
Hello @ifedorenko, 
> This PR is supposed to use little less memory compared to 0.9.5, can't think of why it takes 2x memory instead.

Perhaps my assertion of 2X memory was premature: it used 2x the memory *in the first 15m*, but that could have been because it just progressed farther in that interval than standard 0.9.5. 

I'm monitoring it as it runs, and it seems to have plateaued at about the same 36GB of RAM it reached in those first 15 mins. If it has really plateaued, it would be using only 90% (ie, 10% less) than the ~40GB that standard 0.9.5 ended up using up to the point I had to kill it. 

I will continue to monitor it and report back on its memory use.  

> I'd need to know what data structures use the memory in your tests to suggest anything specific.
Would the output of `curl http://localhost:6060/debug/pprof/heap` provide what you need? Please let me know so I can be prepared. 

Cheers, 
-- Durval.

DurvalMenezes commented:
> I will continue to monitor it and report back on its memory use.
> 
The current `out-of-order-restore-no-progress` restore just blew past the point where I had to kill the previous one (at ~59GB restored) and the RSS memory usage stays at ~36GB, so my initial claim of 2x the memory usage was indeed premature: it seems this PR actually uses ~10% *less* memory than standard restic.

Sorry for the false alarm.

DurvalMenezes commented:
The restore finished a few hours later and, in a word: WOW! :-)

In more words: it finished in 4h08m the same restore that standard 0.9.5 restic took over 72 hours and reached only 2/3 -- so, performance-wise, congratulations @ifedorenko! Thanks to your work, I escaped having to junk restic and all the hard work I invested implementing and tuning it here -- it would not be viable to keep using a backup method that was so slow when restoring things. 

On a less positive note, I detected two corrupted files -- more details on [my forum post](https://forum.restic.net/t/restic-restore-speeds-asymptotically-dropping-towards-zero/1756/8). To sum it up, I think there's a good chance it wasn't caused by restic -- I will do some more tests to try and locate the issue, and then come back to report.

Cheers, 
-- Durval.

DurvalMenezes commented:
> RSS memory usage stays at ~36GB, so my initial claim of 2x the memory usage was indeed premature: it seems this PR actually uses ~10% _less_ memory than standard restic, as @ifedorenko expected.

More info memory-wise: [when I re-run the exact same `restic restore` command, just cd'ing to a new directory (effectivelly changing the restore destination, as I explicitly use `--target=.`)](https://forum.restic.net/t/restic-restore-speeds-asymptotically-dropping-towards-zero/1756/12?u=durval), the maximum RSS usage was only ~24GB, ie, ~33% less than the first time I ran that same command! Does this make any sense to you?

saikpr commented on internal/restorer/doc.go:
```diff
@@ -5,29 +5,20 @@
+// Implementation does not guarantee order in which blobs are written to the
+// target files and, for example, the last blob of a file can be written to the
+// file before any of the preceeding file blobs. It is therefore possible to
+// have gaps in the data written to the target files if restore fails or
```
Not all Filesystems allow multiple open files at different location, would it be possible to accomodate both strategies?

ifedorenko replied:
Not sure I understand the concern. Are you saying some filesystem only support sequential writes to a file? Specific example that demonstrates the problem will be useful here.

saikpr replied:
you can see the different modes in rclone(which comes to my mind), https://rclone.org/commands/rclone_mount/#file-caching

only the `vfs-cache-mode full` will work with this commit, but as of now it works with `vfs-cache-mode minimal` also

ifedorenko replied:
I am afraid 0.9.5 is not compatible with `vfs-cache-mode minimal` already. It may decide to close a file, then reopen it with `O_APPEND` some time later. To support `vfs-cache-mode minimal` we need to go back to pre restic 0.9 impl. I suggest we treat that as a separate feature request.

rawtaz replied:
@ifedorenko Can you possibly reference the commit/PR/change/whatever in 0.9.5 that introduced the incompatibility with `vfs-cache-mode minimal`? Thanks!

ifedorenko replied:
PR #1719 and commit 1869930d95d8c9c1f49e042b5379c0aab72aa353 specifically. 

This is where restore (re)opens files with O_APPEND https://github.com/restic/restic/blob/af200157258579d286d335edef05f0279741602d/internal/restorer/fileswriter.go#L56

rawtaz replied:
We'll leave this for now due to the reasons mentioned by @ifedorenko. Let it be a separate feature request if needed.

saikpr commented on internal/restorer/fileswriter.go:
```diff
@@ -4,98 +4,91 @@ import (
-// of concurrent writeToFile invocations plus cacheCap.
+// writes blobs to target files.
+// multiple files can be written to concurrently.
+// multiple blobs can be concurrently written to the same file.
```
This might cause issues with FUSE based systems, especially FUSE mounted high-latency systems.

rawtaz replied:
@saikpr Can you please elaborate a bit on this? In what way can it cause issues, for what reason in FUSE?

hyperfekt commented:
This code restored my ~60GB backup in half an hour, whereas the old restorer code would have finished in months (no hyperbole) at the rate it was going. In both cases the entire repo was already downloaded. As a user, I'd highly recommend merging this.

saikpr commented:
This did improve the RAM and speed (but not that much if increase the Chunk Size to 128MB)

But Rclone was throwing a lot more  `Didn't finish writing GET request (wrote 135396258/135407930 bytes)` errors than without it.

Eventually both the restores were exactly same, byte-by-byte only the stderr was polluted  with the above error.

You can find the logs here [out_of_band.txt](https://github.com/restic/restic/files/3447991/out_of_band.txt)

ruiztulio commented:
Hi,

I was having an issue with this:

![image](https://user-images.githubusercontent.com/2739530/62550112-4ad5a580-b82f-11e9-94a0-c010bc72d7e4.png)

The last error was because I cancelled the restore process. And also sometimes the restore fail, those backups are about 45G, sometimes the restore only does 35G

But after using the branch from @ifedorenko it worked perfectly:

![image](https://user-images.githubusercontent.com/2739530/62550574-16aeb480-b830-11e9-92b3-21613eccd4c8.png)

As you may notice we are using rest-server (I had to blur url and customer name). Hope this add any value to the discussion.

Regards

saikpr commented:
One quick fix will be to reduce the number of workers (here
https://github.com/restic/restic/blob/master/internal/restorer/filerestorer.go#L24)
or get more RAM :p
--

Sainyam Kapoor

On Tue, Aug 6, 2019 at 10:58 AM Tulio Ruiz <notifications@github.com> wrote:

> Hi,
>
> I was having an issue with this:
>
> [image: image]
> <https://user-images.githubusercontent.com/2739530/62550112-4ad5a580-b82f-11e9-94a0-c010bc72d7e4.png>
>
> The last error was because I cancelled the restore process. And also
> sometimes the restore fail, those backups are about 45G, sometimes the
> restore only does 35G
>
> But after using the branch from @ifedorenko
> <https://github.com/ifedorenko> it worked perfectly:
>
> [image: image]
> <https://user-images.githubusercontent.com/2739530/62550574-16aeb480-b830-11e9-92b3-21613eccd4c8.png>
>
> As you may notice we are using rest-server (I had to blur url and customer
> name). Hope this add any value to the discussion.
>
> Regards
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/restic/restic/pull/2195?email_source=notifications&email_token=AA6K6YLXAWNT7AHKTN7DIPLQDGGSDA5CNFSM4G3K4YYKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3VNRBY#issuecomment-518707335>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA6K6YLOANEYR7Y4BBUFH6LQDGGSDANCNFSM4G3K4YYA>
> .
>

Church- commented:
Having already demo'd this branch at work, I definitely think it needs to be merged in.

Allowed me to circumvent a ton of `out of cache capacity` errors I was running into.

ifedorenko committed:
restorer: Allow writing target file blobs out of order

Much simpler implementation that guarantees each required pack
is downloaded only once (and hence does not need to manage
pack cache). Also improves large file restore performance.

Signed-off-by: Igor Fedorenko <igor@ifedorenko.com>

ifedorenko committed:
restorer: go mod vendor; go mod tidy

Signed-off-by: Igor Fedorenko <igor@ifedorenko.com>

ifedorenko commented:
fyi, rebased my branch on the current master, also split `vendor/...` changes to a separate commit

bkmeneguello commented:
Really waiting this become production to start using restic in my company. We have some huge VM disks to synchronize to GDrive and actually they don't perform well to restore. I've tested this branch and worked perfectly, but until become stable I can't use in production.

StrikeLines commented:
This branch is an absolute life saver. I had to use PR2195 in order to successfully restore about 50 TB of business data. The stock restic restore had been running for weeks with no end in sight.  This version restored **9 TB** of files _overnight!_ 

imo, in it's current state, restic-master is not a usable backup solution. It's terrible restore performance is unacceptable, making it useless on large repositories. 

This pull request should be merged to master as soon as possible. This deserves to be prioritized above all other pending requests. 

Thanks ifedorenko. This saved my bacon.

ifedorenko commented:
For the record, I wrote the current broken implementation too, so I am to blame for the current state of affairs. There is O(N^2) loop there, which gets progressively slower with number of blobs in a file. It can be fixed, but the fix would add complexity to already complex code and I think this PR is overall better deal.

seertenedos commented:
Thanks @ifedorenko this fixed my issue even with a small repository taking over 2 hours to now taking less than 5 minutes. i thought i would have to move away from restic but this makes it usable.  Any idea when this will be in master?

ifedorenko commented:
@seertenedos I don't know when (or if) this will be merged.

seertenedos commented:
@fd0 Are you likely to allow this to be merged in any time soon? It would be fantastic Xmas present for Restic users based on mine and others restore experiences.

robvalca commented:
Just another bump for this to be merged, I think that is crucial to have this on restic master, otherwise restore of big files is unpractical  @fdo  @rawtaz  @saikpr   Thanks !

rawtaz commented on internal/restorer/filerestorer_test.go:
```diff
@@ -56,10 +52,10 @@ func (i *TestRepo) packID(name string) restic.ID {
+// func (i *TestRepo) pack(queue *packQueue, name string) *packInfo {
+// 	id := i.packsNameToID[name]
+// 	return queue.packs[id]
+// }
```
Should we remove this part of dead code to keep it all clean? Or is there some specific reason to keep it?

rawtaz commented on internal/restorer/fileswriter.go:
```diff
@@ -4,98 +4,91 @@ import (
-	n, err := wr.Write(blob)
-	cacheOrCloseWriter(wr)
+
+	// fmt.Printf("write %s offset %d len %d '%s'\n", file.location, offset, len(data), string(data))
```
Should this line be kept?

rawtaz commented on internal/restorer/fileswriter.go:
```diff
@@ -4,98 +4,91 @@ import (
+	_, err = wr.WriteAt(blob, offset)
+
 	if err != nil {
+		// fmt.Printf("error %s %v\n", file.location, err)
```
Should this line be kept?

rawtaz commented:
@ifedorenko What's the difference between this branch and your other branch named `out-order-restore`?

Out of interest I just rebased a local copy of this PR onto latest master (af200157) and there were no conflicts. I added some comments/questions to the changed files though.

ifedorenko commented:
> What's the difference between this branch and your other branch named out-order-restore?

`out-order-restore` has fancy and now abandoned progress UI.

ifedorenko commented:
@rawtaz fyi, I rebased on the latest master and removed commented out debug code as you requested. thank you for your feedback. let me know if you have any other questions.

rawtaz commented:
Thanks @ifedorenko. Besides a couple of eyes from @fd0, we need a changelog file as well, can you add that?

ifedorenko commented:
> Besides a couple of eyes from @fd0, we need a changelog file as well, can you add that?

done

schmitch commented:
@rawtaz what's the state of that? it's really akward that #2244 is happening

eigood commented:
I'm thinking I'd like to voice my opinion on this branch as well.  We are only just getting started with using restic, and are already seeing very poor restore performance.  Our wasabi-based repo has a mere 1070 snapshots, 7 for the primary mainline $host, and then for each of the 177 individual wordpress sites, there are 6.  All contain the same data, various versions of wordpress, going, so lots of chances for dedup on text files.  Input size was originally 120G, and  the first backups stored only 55G after deduping(a nice win).

But now I'm attempting to restore the mainline base host, which will have that 120G. It has ran overnight, and only managed 50G.  It would take 8 hours if I just did an rsync from the existing machine.  I'm running ifstat right now, and I don't see any large network downloads.  Strace shows lots of nanosleep, and other busy work happening, with very little read() from the network, and write() to the disk.

I was hoping I wouldn't have to try this branch; we have a general rule of sticking with only officially released software(less maintenance work on our part).

ps: total repo size is only 96G right now, this tiny compared to others I've seen mentioned.

eigood commented:
I'm thinking I'd like to voice my opinion on this branch as well.  We are only just getting started with using restic, and are already seeing very poor restore performance.  Our wasabi-based repo has a mere 1070 snapshots, 7 for the primary mainline $host, and then for each of the 177 individual wordpress sites, there are 6.  All contain the same data, various versions of wordpress, going, so lots of chances for dedup on text files.  Input size was originally 120G, and  the first backups stored only 55G after deduping(a nice win).

But now I'm attempting to restore the mainline base host, which will have that 120G. It has ran overnight, and only managed 50G.  It would take 8 hours if I just did an rsync from the existing machine.  I'm running ifstat right now, and I don't see any large network downloads.  Strace shows lots of nanosleep, and other busy work happening, with very little read() from the network, and write() to the disk.

I was hoping I wouldn't have to try this branch; we have a general rule of sticking with only officially released software(less maintenance work on our part).

ps: total repo size is only 96G right now, this tiny compared to others I've seen mentioned.

rawtaz commented:
@schmitch You mean the status of *this* PR? I'm waiting for some eyes on it from @fd0, and also @ifedorenko should probably check the three (or so) comments I made on lines in the code :)

ifedorenko commented:
> @ifedorenko should probably check the three (or so) comments I made on lines in the code :)

@rawtaz I don't see any unanswered comments. can you be more specific what I missed?

dimejo commented:
> I don't see any unanswered comments. can you be more specific what I missed?

He's probably referring to the comments which were addressed by commit 430e058.

rawtaz commented:
@ifedorenko I'm terribly sorry for saying there were unresolved lines. You indeed fixed that a while ago. GitHub for some reason still showed them as unresolved, so I thought they weren't. I clicked the resolve button now, so no more confusion on that part :P

I took the liberty of updating the changelog file a bit and squashed some commits. From where I stand this PR is ready for merging, just going to get @fd0 on it in a short while before we push the button. I have no doubts it's going to be merged soon.

ifedorenko commented:
@rawtaz I would prefer two commits, to be honest, one for the code changes including changelog, and the other for `go mod ...` generated noise.

fd0 reviewed (approved):
Awesome work, thank you very much @ifedorenko! And I'm very sorry it took us so long to get this merged :)

buhkguj commented:
@fd0: We have the same problem. Restoring from 1TB repository brings up the caching errors on serveral files. We are using 0.9.6. Please release this fix ASAP if not done already!

rawtaz commented:
> We have the same problem. Restoring from 1TB repository brings up the caching errors on serveral files. We are using 0.9.6.

@buhkguj Please be more specific. The same problem as *what*? Is the problem you are referring to already documented somewhere? If so, please provide a link to it. Thanks.

buhkguj commented:
> @buhkguj Please be more specific. The same problem as _what_? Is the problem you are referring to already documented somewhere? If so, please provide a link to it. Thanks.

We got restore errors like this: `"not enough cache capacity: requested 9976743, available 6836375"`

Issue: https://github.com/restic/restic/issues/2244
This PR shoud fix, but I think it is not included in latest release yet. (0.9.6)
Thanks for your great work.

rawtaz commented:
@buhkguj You're correct in that this PR is not included in 0.9.6 (it will be in the next release, of course). If you want to try this PR, you can download the latest master build from [here](https://beta.restic.net/?sort=time&order=desc). If you do, please let us know if it helps with the problems you're having!

i0x71 commented:
Guys,
Its been 11 months since this fix was made available, the version available on the main site this doesn't Include it. This really brings down the morale and trust in the team, I would imagine there are lots of people out there who faced this issue (me included) only to find themselves on this page. 11 months the after fix was available.

saikpr commented:
@i0x71 @rawtaz You can find the binaries on the beta website here : https://beta.restic.net/restic-v0.9.6-160-gf033850a/
I am pretty sure most of the maintainers have been busy given the present situation, and they would have a lot on their mind.

rawtaz commented:
@i0x71 That is 11 months during which you and everyone else who had this specific problem has had all the opportunity in the world to try/use this fix, and in fact some users who wanted to try it actually did so and have been using it ever since.

All you had to do was build restic yourself (which is just one command assuming you have Go installed), and if you didn't know about this you just had to ask. We'd be happy to help you in case you didn't know how to do it.

That said, can you tell me in what way the team that maintains restic is less trustworthy due to it taking a long time to get this fix into master (and released)? In what way does us having a life to manage, sometimes with severe issues to deal with that you cannot even imagine, and having kids, family, relatives and a full time job including travels to work out during the approximately 15 hours we're awake, make us less trustworthy?

On the contrary I would suggest that we'd be less trustworthy if we merged this and furthermore put it out as an official release the day after the PR was created. I hope you understand why that would be more of an issue.

FWIW, there will always be changes and fixes that are either waiting to be merged or that have been merged but not put into a release, this was just one of them. I agree it took too long, but there are reasons for that and anyone who needed this fix had the possibility to use it, you don't need a release for that. At the very least just ask if you can try it beforehand. Thanks!

buhkguj commented:
@rawtaz: I tested the beta bin and the restore was done without errors!
Thanks for the tip. I realy understand that it takes time, specially during this strange period we all live in right now. Anyway I hope there is a way to release it soon, so we can use it on production.
Thanks again for this great tool!

julijane commented:
I'm sorry but I sort of have to agree with the criticism. Today I had to panic-restore some data and then stumbled across the errors and that really did not lighten my day as I it increased my panic even more. It took me some googling and clicking through issues here to find this and the link to the beta binaries which then could properly restore the data.

While I understand some of the reasoning I really think that if there is an issue with the released version not restoring (all) data it is a good idea to release the fixed version in time. People in my situation will clearly be thankful because when you are in need of a restore it really does not lighten the day when the restore fails and/or takes ages.

Just my 2ct, thanks for considering my thoughts.

dhilgarth commented:
It might help if there would be some kind of roadmap where users would be able to see when the new version will be released.
Does something like that exist?

darkdragon-001 commented:
@dhilgarth One it's released, you will see the release tag for the commits in this pull request ([StackOverflow](https://stackoverflow.com/a/47286371/3779655)). In addition, some [roadmap](https://github.com/features/project-management) to see which issues are blocking a release would be nice.

rawtaz commented:
> Does something like that exist?

No. And as you can see there's a huge flow of issues and comments taking up a lot of time for the maintainers. It's hard to manage all that and we need the flexibility in releasing when we think that we are able to. Having to maintain a roadmap would in my personal opinion only be counter-productive, it would not help.

darkdragon-001 commented:
@rawtaz All these project management tools help to stay focused and solve the important issues first instead of being overwhelmed by the huge flow of issues. Overall productivity will increase. If it was my project, I would make use of them. But the team probably knows best what works for them...

rawtaz commented:
I spent some time with one of them and it turned to not work at all. It was a complete waste of time, and I was honestly a bit annoyed by the fact that it advertises itself as this great thing, yet when you try it it's utter garbage. I guess this hasn't helped raising my love for such tools :D To be fair, it wasn't a project management tool specifically, more like a GitHub issues+PR management tool. Anyway, I hear you, and my comment was only about how I feel about it, perhaps others in the team want to or use such tools. Regardless, I truly don't think having a roadmap would help us. A release will come when it's ready.

darkdragon-001 commented:
@rawtaz To get me right, I am not requesting some sort of timeline _when_ to expect the next release. It would be more to assign issues to milestones (which represent the next minor, major and future releases). This way, users can see which issues are blocking a release and estimate how much work is left before a release can happen. Using such tools allows to separate _communication on (new) issues_ and _working on / fixing issues_. This way one can stay focused while working on issues which usually increases productivity a lot.

darkdragon-001 commented:
The [dotnet](https://github.com/dotnet/runtime/projects) team uses them extensively to organize the 5000+ open issues. Other examples for inspiration: [docker](https://github.com/docker/roadmap/projects/1), [mono](https://github.com/mono/mono/projects/9), [OpenZFS](https://github.com/openzfs/zfs/projects/25).

rawtaz commented:
That's great. I am not going to add any further comments on this. Thanks for your suggestion!
